PC/winreg.c

```diff
@@ -1488,53 +1502,76 @@ static PyObject *
 winreg_QueryValue_impl(PyObject *module, HKEY key, const Py_UNICODE *sub_key)
 /*[clinic end generated code: output=c655810ae50c63a9 input=41cafbbf423b21d6]*/
 {
-    long rc;
-    PyObject *retStr;
-    wchar_t *retBuf;
-    DWORD bufSize = 0;
-    DWORD retSize = 0;
-    wchar_t *tmp;
+    LONG rc;
+    HKEY childKey = key;
+    WCHAR buf[256], *pbuf = buf;
+    DWORD size = sizeof(buf);
+    DWORD type;
+    Py_ssize_t length;
+    PyObject *result = NULL;
```

Comment:
Can we rewrite this as a wrapper around the `QueryValueEx` implementation? It should only require adding an extra argument and splitting out the result.

Reply:
Doesn't matter if the audit event is raised twice.

Reply:
We'd still have to reject `HKEY_PERFORMANCE_DATA` here, and we'd still have to open `sub_key` if it's a non-empty string. 

I think the loop in `winreg_QueryValueEx_impl()` should be updated to match what I wrote here. Use an initial stack buffer that's big enough for common cases, so we only need one system call in most cases. If needed, realloc a buffer to exactly the returned size instead of playing a guessing game with `bufSize *= 2`. Use a clean exit label instead scattering `PyMem_Free()` around the code at arbitrary exit points.

Reply:
Those all sound like good changes, but we may be outgrowing the scope here. Maybe just go with `NotImplementedError` when the APIs aren't available and file a new bug to improve both implementations?

Reply:
I guess since those are somewhat unrelated they can go in first. I will open an issue + PR for this. In case that helps with the review I can split off the general "cleanup" done here e.g. to remove code only relevant until WIndows 7. This would reduce the size of this PR by quite a bit.

Reply:
If you're happy to split up the Win7 cleanup, I'm fine with that. There'll be a `WINVER` definition somewhere that can be updated as well - I forget exactly where we keep that.

Or I'm also okay with deferring the registry function improvements. We don't officially support the app or games families right now (the Store app is full trust, so regular desktop), so we don't strictly have to include all existing functions.

Reply:
https://github.com/python/cpython/pull/102337 pulls out the Windows 7 cleanup from this PR.

Reply:
`WINVER` is apparently already updated: https://github.com/python/cpython/blob/f300a1fa4c121f7807cbda4fc8bb26240c69ea74/PC/pyconfig.h#L147

Reply:
#102345 for the winreg changes